File: ojuz/ioi/2020/Day1/tickets_sol/solution/wy-subtask5.cpp
#include "tickets.h"
#include <bits/stdc++.h>
using namespace std;

int sum[1505], f[1505], b[1505];
long long dp[85][3505];
int best[85][3505];
vector< vector<int> > ret;

long long find_maximum(int k, std::vector<std::vector<int>> d) {
	int c = d.size();
	int s = d[0].size();
	vector< pair<long long, pair<int, int> > > V;
	long long ans = 0;
	vector<vector<pair<int, int>>> d2;
	d2.resize(c);
	for (int i = 0; i < c; i++) {
		for (int j = 0; j < s; j++) {
			d2[i].push_back(make_pair(d[i][j], j));
		}
		sort(d2[i].begin(), d2[i].end());
	}

	memset(dp, -63, sizeof(dp));
	dp[0][0] = 0;
	for (int i = 1; i <= c; i++) {
		for (int j = 0; j <= c*k/2; j++) {
			long long minus = 0, plus = 0;
			for (int l = 0; l < k; l++) minus += d2[i-1][l].first;
			for (int l = 0; l <= min(j, k); l++) {
				long long nans = dp[i-1][j-l] + plus - minus;
				if (nans > dp[i][j]) {
					dp[i][j] = nans;
					best[i][j] = j-l;
				}
				if (l != min(j, k)) {
					minus -= d2[i-1][k-l-1].first;
					plus += d2[i-1][s-l-1].first;
				}
			}
		}
	}

	int cur = c*k/2;
	for (int i = c-1; i >= 0; i--) {
		sum[i] = cur - best[i+1][cur];
		cur = best[i+1][cur];
	}


	ret.resize(c);
	for (int i = 0; i < c; i++) {
		vector<int> v;
		v.resize(s, -1);
		ret[i] = v;
	}
	for (int i = 0; i < k; i++) {
		vector< pair<int, int> > v;
		for (int j = 0; j < c; j++) v.push_back(make_pair(sum[j], j));
		sort(v.begin(), v.end());
		for (int j = 0; j < c/2; j++) {
			ret[v[j].second][d2[v[j].second][f[v[j].second]].second] = i;
			f[v[j].second]++;
		}
		for (int j = c/2; j < c; j++) {
			ret[v[j].second][d2[v[j].second][s-b[v[j].second]-1].second] = i;
			b[v[j].second]++;
			sum[v[j].second]--;
		}
	}
	allocate_tickets(ret);
	return dp[c][c*k/2];
}

File: ojuz/ioi/2020/Day1/tickets_sol/solution/wy-subtask5.java
class tickets {
	long find_maximum(int k, int[][] d) {
		int c = d.length;
		int s = d[0].length;
		long[][] dp = new long[c][c*k+1];
		long[][] rewards = new long[c][k+1];
		for(int i=0; i<c; i++) {
			rewards[i][0] = 0;
			for(int j=0; j<k; j++) {
				rewards[i][0] -= d[i][j];
			}
			for(int j=1; j<=k; j++) {
				rewards[i][j] = rewards[i][j-1] + d[i][s-j]
								+ d[i][k-j];
			}
		}
		for(int i=0; i<c; i++) {
			for(int l=0; l<=c*k; l++) {
				dp[i][l] = (long)(-1e18);
			}
		}
		int[][] from = new int[c][c*k/2+1];
		for(int j=0; j<=k; j++) {
			dp[0][j] = rewards[0][j];
			from[0][j] = j;
		}
		for(int i=1; i<c; i++) {
			for(int j=0; j<=k; j++) {
				for(int l=0; l+j<=c*k/2; l++) {
					if(dp[i-1][l] + rewards[i][j] > dp[i][l+j]) {
						dp[i][l+j] = dp[i-1][l] + rewards[i][j];
						from[i][l+j] = j;
					}
				}
			}
		}
		int[][] ret = new int[c][s];
		for(int i=0; i<c; i++) {
			for(int j=0; j<s; j++) {
				ret[i][j] = -1;
			}
		}
		int current = c*k/2; // the dp state
		int minus_day = 0;
		int plus_day = c*k-1;
		for(int i=c-1; i>=0; i--) {
			for(int j=0; j<k-from[i][current]; j++) {
				ret[i][j] = minus_day % k;
				minus_day++;
			}
			for(int j=s-from[i][current]; j<s; j++) {
				ret[i][j] = plus_day % k;
				plus_day--;
			}
			current -= from[i][current];
		}
		grader.allocate_tickets(ret);
		return dp[c-1][c*k/2];
	}
}
